src/MemoryTracing/MemoryTracingIntegrationTest.cpp

```diff
@@ -0,0 +1,129 @@
+ public:
+  MemoryTracingIntegrationTestFixture(uint64_t memory_sampling_period_ns)
+      : memory_sampling_period_ns_(memory_sampling_period_ns) {}
+
+  void StartTracingAndWaitForTracingLoopStarted() {
+    CHECK(!producer_.has_value());
+    CHECK(!listener_.has_value());
+    producer_.emplace(memory_sampling_period_ns_);
+    listener_.emplace();
+    producer_->SetListener(&*listener_);
+    producer_->Start();
+  }
+
+  [[nodiscard]] std::vector<ProducerCaptureEvent> StopTracingAndGetEvents() {
+    CHECK(producer_.has_value());
+    CHECK(listener_.has_value());
+    producer_->Stop();
+    producer_.reset();
+    std::vector<ProducerCaptureEvent> events = listener_->GetAndClearEvents();
+    listener_.reset();
+    return events;
+  }
+
+ private:
+  uint64_t memory_sampling_period_ns_;
+  std::optional<MemoryInfoProducer> producer_ = std::nullopt;
+  std::optional<BufferMemoryInfoListener> listener_ = std::nullopt;
+};
+
+[[nodiscard]] std::vector<ProducerCaptureEvent> TraceAndGetEvents(
+    MemoryTracingIntegrationTestFixture* fixture) {
+  CHECK(fixture != nullptr);
+
+  fixture->StartTracingAndWaitForTracingLoopStarted();
+
+  constexpr absl::Duration kSleepBeforeStopTracing = absl::Nanoseconds(1000000000);
+  absl::SleepFor(kSleepBeforeStopTracing);
+  return fixture->StopTracingAndGetEvents();
+}
+
+void VerifyOrderAndSamplingPeriodOfEvents(const std::vector<ProducerCaptureEvent>& events,
+                                          uint64_t sampling_period_ns) {
+  constexpr uint64_t kSamplingTimeErrorToleranceNs = 1000000;
+  uint64_t previous_event_timestamp_ns = 0;
+  for (const auto& event : events) {
+    if (event.event_case() != ProducerCaptureEvent::kSystemMemoryUsage) UNREACHABLE();
+
+    uint64_t current_event_timestamp_ns = event.system_memory_usage().timestamp_ns();
+    EXPECT_GE(current_event_timestamp_ns, previous_event_timestamp_ns);
+
+    if (previous_event_timestamp_ns > 0 && sampling_period_ns > 0) {
+      uint64_t time_since_last_successful_sampling =
+          current_event_timestamp_ns - previous_event_timestamp_ns;
+      uint64_t periods_since_last_successful_sampling =
+          time_since_last_successful_sampling / sampling_period_ns;
+      // If there are failed attempts of memory sampling, the sampling time error would be
+      // difficult to decide. Hence, we only check the case of successive successful attempts of
+      // sampling.
+      if (periods_since_last_successful_sampling <= 1) {
+        EXPECT_GE(time_since_last_successful_sampling,
+                  sampling_period_ns - kSamplingTimeErrorToleranceNs);
+        EXPECT_LE(time_since_last_successful_sampling,
+                  sampling_period_ns + kSamplingTimeErrorToleranceNs);
+      }
+    }
+
+    previous_event_timestamp_ns = current_event_timestamp_ns;
+  }
+}
+
+}  // namespace
+
+TEST(MemoryTracingIntegrationTest, SystemMemoryUsageTracing) {
+  constexpr uint64_t kMemorySamplingPeriodNs = 100000000;
```

Comment:
nitpick: you can use "'" to separate groups of 000 so that it is easier to read: as in `100'000'000`. Or better yet use `absl::Milliseconds(100) / absl::Nanoseconds(1)`

Reply:
Changed to use `absl::Milliseconds(100) / absl::Nanoseconds(1)`. Thanks for the suggestion!